File: internal/engine/snapshot.go
package engine

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// SnapshotOptions controls the deterministic accessibility-tree renderer.
// Depth is measured from the selected root (root is depth zero); zero means no
// limit. RootNodeID is populated internally after resolving --selector.
type SnapshotOptions struct {
	Interactive bool   `json:"interactive,omitempty"`
	Compact     bool   `json:"compact,omitempty"`
	Depth       int    `json:"depth,omitempty"`
	Selector    string `json:"selector,omitempty"`
	URLs        bool   `json:"urls,omitempty"`
	Diff        bool   `json:"diff,omitempty"`
	Since       string `json:"since,omitempty"`
	RootNodeID  string `json:"-"`
}

// SnapshotResult is the stable output shape for snapshot --json.
type SnapshotResult struct {
	SnapshotID string                 `json:"snapshot_id,omitempty"`
	Tree       string                 `json:"tree"`
	Refs       map[string]SnapshotRef `json:"refs"`
	Hint       string                 `json:"hint,omitempty"`
}

// SnapshotRef describes the protocol-neutral target behind one rendered ref.
type SnapshotRef struct {
	NodeID         string            `json:"node_id"`
	BackendNodeID  int64             `json:"backend_node_id,omitempty"`
	Role           string            `json:"role"`
	Name           string            `json:"name,omitempty"`
	Value          string            `json:"value,omitempty"`
	State          string            `json:"state,omitempty"`
	Visible        bool              `json:"visible"`
	Interactive    bool              `json:"interactive"`
	URL            string            `json:"url,omitempty"`
	RefKey         string            `json:"refkey,omitempty"`
	DOMPath        string            `json:"dom_path,omitempty"`
	SiblingOrdinal int               `json:"sibling_ordinal,omitempty"`
	Attributes     map[string]string `json:"attributes,omitempty"`
}

// AXSelectorResolver resolves a CSS selector to an accessibility node's
// backend DOM node id. It is optional so renderer tests can inject AX nodes
// without a browser or CDP implementation.
type AXSelectorResolver interface {
	AXNodeForSelector(context.Context, Page, string) (string, error)
}

// Snapshot retrieves and renders the current page accessibility tree.
func (s *NavigationService) Snapshot(ctx context.Context, options SnapshotOptions) (SnapshotResult, error) {
	s.snapshotMu.Lock()
	defer s.snapshotMu.Unlock()
	return s.captureSnapshot(ctx, options)
}

// RenderSnapshot renders protocol-neutral AXNode payloads into deterministic
// text. Stable session-local ref allocation is applied by NavigationService.
func RenderSnapshot(nodes []AXNode, options SnapshotOptions) (SnapshotResult, error) {
	if options.Depth < 0 {
		return SnapshotResult{}, errors.New("snapshot depth cannot be negative")
	}
	parsed := make(map[string]*snapshotNode, len(nodes))
	order := make([]string, 0, len(nodes))
	for index, node := range nodes {
		parsedNode, err := decodeSnapshotNode(node.Raw, index)
		if err != nil {
			return SnapshotResult{}, fmt.Errorf("decode accessibility node %d: %w", index, err)
		}
		if _, exists := parsed[parsedNode.id]; exists {
			return SnapshotResult{}, fmt.Errorf("duplicate accessibility node id %q", parsedNode.id)
		}
		parsed[parsedNode.id] = parsedNode
		order = append(order, parsedNode.id)
	}
	if len(parsed) == 0 {
		return SnapshotResult{Tree: "", Refs: map[string]SnapshotRef{}}, nil
	}

	children := make(map[string][]string, len(parsed))
	childOf := make(map[string]bool, len(parsed))
	for _, id := range order {
		node := parsed[id]
		for _, childID := range node.childIDs {
			if _, exists := parsed[childID]; !exists || childID == id {
				continue
			}
			children[id] = append(children[id], childID)
			childOf[childID] = true
		}
	}
	// Older or injected payloads may only carry parentId. Add those links while
	// preserving the input order of the AX tree.
	for _, id := range order {
		node := parsed[id]
		if node.parentID != "" && node.parentID != id && parsed[node.parentID] != nil && !contains(children[node.parentID], id) {
			children[node.parentID] = append(children[node.parentID], id)
			childOf[id] = true
		}
	}
	roots := make([]string, 0, len(parsed))
	if options.RootNodeID != "" {
		if parsed[options.RootNodeID] == nil {
			return SnapshotResult{}, fmt.Errorf("snapshot selector resolved to unknown accessibility node %q", options.RootNodeID)
		}
		roots = append(roots, options.RootNodeID)
	} else {
		for _, id := range order {
			if !childOf[id] {
				roots = append(roots, id)
			}
		}
		if len(roots) == 0 {
			roots = append(roots, order...)
		}
	}
	sort.SliceStable(roots, func(i, j int) bool { return snapshotIDLess(roots[i], roots[j]) })

	for _, node := range parsed {
		node.children = children[node.id]
	}
	assignSnapshotPaths(parsed, roots)
	refs := make(map[string]SnapshotRef)
	lines := make([]string, 0, len(parsed))
	visited := make(map[string]bool, len(parsed))
	refNumber := 0
	for _, rootID := range roots {
		lines = append(lines, renderSnapshotNode(parsed, rootID, options, 0, visited, &refNumber, refs)...)
	}
	return SnapshotResult{Tree: strings.Join(lines, "\n"), Refs: refs}, nil
}

type snapshotNode struct {
	id             string
	parentID       string
	role           string
	name           string
	description    string
	value          string
	url            string
	ignored        bool
	interactive    bool
	iframe         bool
	shadowRoot     bool
	backendNodeID  int64
	state          map[string]string
	attributes     map[string]string
	visible        bool
	childIDs       []string
	children       []string
	domPath        string
	siblingOrdinal int
	detached       bool
}

func decodeSnapshotNode(raw json.RawMessage, index int) (*snapshotNode, error) {
	if len(raw) == 0 || string(raw) == "null" {
		return nil, errors.New("empty node payload")
	}
	var payload struct {
		NodeID         json.RawMessage   `json:"nodeId"`
		ParentID       json.RawMessage   `json:"parentId"`
		Role           json.RawMessage   `json:"role"`
		Name           json.RawMessage   `json:"name"`
		Description    json.RawMessage   `json:"description"`
		Value          json.RawMessage   `json:"value"`
		Ignored        bool              `json:"ignored"`
		BackendNodeID  json.RawMessage   `json:"backendDOMNodeId"`
		ChildIDs       []json.RawMessage `json:"childIds"`
		FrameID        string            `json:"frameId"`
		IsFrameOwner   bool              `json:"isFrameOwner"`
		IsShadowRoot   bool              `json:"isShadowRoot"`
		ShadowRoot     bool              `json:"shadowRoot"`
		ShadowBoundary bool              `json:"shadowBoundary"`
		Detached       bool              `json:"detached"`
		IsDetached     bool              `json:"isDetached"`
		Properties     []struct {
			Name  string          `json:"name"`
			Value json.RawMessage `json:"value"`
		} `json:"properties"`
		Visible     *bool             `json:"visible"`
		Attributes  map[string]string `json:"attributes"`
		Alt         string            `json:"alt"`
		Title       string            `json:"title"`
		Placeholder string            `json:"placeholder"`
		TestID      string            `json:"testid"`
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, err
	}
	id := rawStringID(payload.NodeID)
	if id == "" {
		id = fmt.Sprintf("index-%d", index)
	}
	role := normalizeSnapshotRole(rawValue(payload.Role))
	name := rawValue(payload.Name)
	value := rawValue(payload.Value)
	node := &snapshotNode{
		id:            id,
		parentID:      rawStringID(payload.ParentID),
		role:          role,
		name:          name,
		description:   rawValue(payload.Description),
		value:         value,
		ignored:       payload.Ignored,
		backendNodeID: rawInt64(payload.BackendNodeID),
		state:         make(map[string]string),
		visible:       !payload.Ignored,
		attributes:    make(map[string]string),
		childIDs:      make([]string, 0, len(payload.ChildIDs)),
		detached:      payload.Detached || payload.IsDetached,
		iframe:        payload.IsFrameOwner || strings.Contains(role, "iframe") || role == "frame" || strings.Contains(strings.ToLower(rawValue(payload.Role)), "frame"),
		shadowRoot:    payload.IsShadowRoot || payload.ShadowRoot || payload.ShadowBoundary || strings.Contains(role, "shadow"),
	}
	if payload.FrameID != "" && (role == "iframe" || role == "frame") {
		node.iframe = true
	}
	for _, childID := range payload.ChildIDs {
		if id := rawStringID(childID); id != "" {
			node.childIDs = append(node.childIDs, id)
		}
	}
	if payload.Visible != nil {
		node.visible = *payload.Visible
	}
	for key, value := range payload.Attributes {
		node.attributes[strings.ToLower(strings.TrimSpace(key))] = value
	}
	for key, value := range map[string]string{"alt": payload.Alt, "title": payload.Title, "placeholder": payload.Placeholder, "testid": payload.TestID} {
		if value != "" {
			node.attributes[key] = value
		}
	}
	for _, property := range payload.Properties {
		propertyName := strings.ToLower(strings.TrimSpace(property.Name))
		propertyValue := rawValue(property.Value)
		switch propertyName {
		case "url", "href":
			if node.url == "" {
				node.url = propertyValue
			}
		case "focusable", "editable", "checked", "expanded", "selected", "disabled", "pressed":
			if propertyName != "focusable" && propertyName != "editable" || propertyBoolean(property.Value) {
				node.state[propertyName] = propertyValue
			}
			if propertyBoolean(property.Value) && propertyName != "expanded" {
				node.interactive = true
			}
		case "type", "autocomplete":
			if propertyValue != "" {
				node.attributes[propertyName] = propertyValue
			}
		}
	}
	if node.role == "link" && node.url == "" {
		// Some CDP revisions expose href as a direct field instead of an AX
		// property; a second, intentionally narrow pass handles that shape.
		var direct struct{ URL, Href string }
		_ = json.Unmarshal(raw, &direct)
		if direct.URL != "" {
			node.url = direct.URL
		} else {
			node.url = direct.Href
		}
	}
	node.interactive = node.interactive || interactiveSnapshotRole(node.role)
	return node, nil
}

func renderSnapshotNode(parsed map[string]*snapshotNode, id string, options SnapshotOptions, depth int, visited map[string]bool, refNumber *int, refs map[string]SnapshotRef) []string {
	node := parsed[id]
	if node == nil || visited[id] || (options.Depth > 0 && depth > options.Depth) {
		return nil
	}
	visited[id] = true
	include := !node.ignored && (!options.Interactive || node.interactive || node.iframe || node.shadowRoot) && (!options.Compact || node.interactive || node.iframe || node.shadowRoot)
	childrenDepth := depth
	result := make([]string, 0)
	if include {
		*refNumber++
		ref := fmt.Sprintf("e%d", *refNumber)
		refs[ref] = SnapshotRef{
			NodeID: node.id, BackendNodeID: node.backendNodeID, Role: node.role, Name: node.name,
			Value: node.value, State: snapshotState(node.state), Visible: node.visible, Interactive: node.interactive, URL: node.url, RefKey: RefKey(node.role, node.name, node.domPath, node.siblingOrdinal),
			DOMPath: node.domPath, SiblingOrdinal: node.siblingOrdinal,
			Attributes: cloneStringMap(node.attributes),
		}
		line := strings.Repeat("  ", depth) + "- " + node.role
		if node.name != "" {
			line += " \"" + strings.ReplaceAll(node.name, "\"", "\\\"") + "\""
		}
		if node.value != "" && node.name == "" {
			line += " = \"" + strings.ReplaceAll(node.value, "\"", "\\\"") + "\""
		}
		if node.iframe {
			line += " [iframe]"
		}
		if node.shadowRoot {
			line += " [shadow-root]"
		}
		if options.URLs && node.url != "" {
			line += " (" + node.url + ")"
		}
		line += " [ref=" + ref + "]"
		result = append(result, line)
		childrenDepth++
	}
	for _, childID := range node.children {
		result = append(result, renderSnapshotNode(parsed, childID, options, childrenDepth, visited, refNumber, refs)...)
	}
	return result
}

func rawValue(raw json.RawMessage) string {
	if len(raw) == 0 || string(raw) == "null" {
		return ""
	}
	var object struct {
		Value json.RawMessage `json:"value"`
	}
	if raw[0] == '{' && json.Unmarshal(raw, &object) == nil && len(object.Value) > 0 {
		return rawScalar(object.Value)
	}
	return rawScalar(raw)
}

func rawScalar(raw json.RawMessage) string {
	if len(raw) == 0 || string(raw) == "null" {
		return ""
	}
	var text string
	if json.Unmarshal(raw, &text) == nil {
		return text
	}
	var number json.Number
	if json.Unmarshal(raw, &number) == nil {
		return number.String()
	}
	var value any
	if json.Unmarshal(raw, &value) == nil {
		return fmt.Sprint(value)
	}
	return ""
}

func rawStringID(raw json.RawMessage) string { return rawScalar(raw) }

func rawInt64(raw json.RawMessage) int64 {
	value, err := strconv.ParseInt(rawScalar(raw), 10, 64)
	if err != nil {
		return 0
	}
	return value
}

func propertyBoolean(raw json.RawMessage) bool {
	value := strings.EqualFold(rawValue(raw), "true")
	if value {
		return true
	}
	var object struct {
		Value any `json:"value"`
	}
	if json.Unmarshal(raw, &object) == nil {
		return strings.EqualFold(fmt.Sprint(object.Value), "true")
	}
	return false
}

func normalizeSnapshotRole(role string) string {
	role = strings.TrimSpace(role)
	if strings.EqualFold(role, "rootwebarea") || strings.EqualFold(role, "webarea") {
		return "document"
	}
	if role == "" {
		return "generic"
	}
	return strings.ToLower(role)
}

func interactiveSnapshotRole(role string) bool {
	switch role {
	case "button", "checkbox", "combobox", "link", "listbox", "menuitem", "option", "radio", "searchbox", "slider", "spinbutton", "switch", "tab", "textbox", "treeitem", "gridcell":
		return true
	default:
		return false
	}
}

func snapshotIDLess(left, right string) bool {
	leftNumber, leftErr := strconv.ParseInt(left, 10, 64)
	rightNumber, rightErr := strconv.ParseInt(right, 10, 64)
	if leftErr == nil && rightErr == nil && leftNumber != rightNumber {
		return leftNumber < rightNumber
	}
	return left < right
}

func contains(values []string, want string) bool {
	for _, value := range values {
		if value == want {
			return true
		}
	}
	return false
}

File: crates/symbrowse-engine/src/snapshot.rs
use crate::AxNode;
use crate::refs::{normalize_ref_part, ref_key};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::cmp::Ordering;
use std::collections::{BTreeMap, BTreeSet, HashMap};

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct SnapshotOptions {
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub interactive: bool,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub compact: bool,
    #[serde(default, skip_serializing_if = "is_zero_i32")]
    pub depth: i32,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub selector: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub urls: bool,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub diff: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub since: String,
    #[serde(skip)]
    pub root_node_id: String,
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct SnapshotResult {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub snapshot_id: String,
    pub tree: String,
    pub refs: BTreeMap<String, SnapshotRef>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub hint: String,
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct SnapshotRef {
    pub node_id: String,
    #[serde(default, skip_serializing_if = "is_zero_i64")]
    pub backend_node_id: i64,
    pub role: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub value: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub state: String,
    pub visible: bool,
    pub interactive: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub refkey: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub dom_path: String,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub sibling_ordinal: usize,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub attributes: BTreeMap<String, String>,
}

#[derive(Clone, Debug)]
struct SnapshotNode {
    id: String,
    parent_id: String,
    role: String,
    name: String,
    value: String,
    url: String,
    ignored: bool,
    interactive: bool,
    iframe: bool,
    shadow_root: bool,
    backend_node_id: i64,
    state: BTreeMap<String, String>,
    attributes: BTreeMap<String, String>,
    child_ids: Vec<String>,
    children: Vec<String>,
    visible: bool,
    dom_path: String,
    sibling_ordinal: usize,
}

/// Render raw AX payloads into deterministic text and temporary refs. Pass the
/// result through [`crate::refs::StableRefRegistry::apply`] for session refs.
pub fn render_snapshot(
    nodes: &[AxNode],
    options: &SnapshotOptions,
) -> Result<SnapshotResult, String> {
    let mut parsed = BTreeMap::new();
    let mut order = Vec::with_capacity(nodes.len());
    for (index, raw) in nodes.iter().enumerate() {
        let node = decode_node(raw, index)?;
        if parsed.insert(node.id.clone(), node.clone()).is_some() {
            return Err(format!("duplicate accessibility node id {:?}", node.id));
        }
        order.push(node.id);
    }
    if parsed.is_empty() {
        return Ok(SnapshotResult::default());
    }

    let mut children: HashMap<String, Vec<String>> = HashMap::new();
    let mut child_of = BTreeSet::new();
    for id in &order {
        let node = &parsed[id];
        for child_id in &node.child_ids {
            if parsed.contains_key(child_id) && child_id != id {
                children
                    .entry(id.clone())
                    .or_default()
                    .push(child_id.clone());
                child_of.insert(child_id.clone());
            }
        }
    }
    for id in &order {
        let node = &parsed[id];
        if !node.parent_id.is_empty()
            && node.parent_id != *id
            && parsed.contains_key(&node.parent_id)
            && !children
                .get(&node.parent_id)
                .is_some_and(|items| items.iter().any(|item| item == id))
        {
            children
                .entry(node.parent_id.clone())
                .or_default()
                .push(id.clone());
            child_of.insert(id.clone());
        }
    }
    for (id, node) in &mut parsed {
        node.children = children.remove(id).unwrap_or_default();
    }

    let mut roots = if !options.root_node_id.is_empty() {
        if !parsed.contains_key(&options.root_node_id) {
            return Err(format!(
                "snapshot selector resolved to unknown accessibility node {:?}",
                options.root_node_id
            ));
        }
        vec![options.root_node_id.clone()]
    } else {
        order
            .iter()
            .filter(|id| !child_of.contains(*id))
            .cloned()
            .collect::<Vec<_>>()
    };
    if roots.is_empty() {
        roots = parsed.keys().cloned().collect();
    }
    roots.sort_by(|left, right| snapshot_id_cmp(left, right));
    assign_paths(&mut parsed, &roots);

    let mut refs = BTreeMap::new();
    let mut lines = Vec::new();
    let mut visited = BTreeSet::new();
    let mut ref_number = 0usize;
    for root in roots {
        render_node(
            &parsed,
            &root,
            options,
            0,
            &mut visited,
            &mut ref_number,
            &mut refs,
            &mut lines,
        );
    }
    Ok(SnapshotResult {
        snapshot_id: String::new(),
        tree: lines.join("\n"),
        refs,
        hint: String::new(),
    })
}

fn decode_node(raw: &Value, index: usize) -> Result<SnapshotNode, String> {
    let object = raw
        .as_object()
        .ok_or_else(|| format!("accessibility node {index} is not an object"))?;
    let id = raw_string_id(object.get("nodeId")).unwrap_or_else(|| format!("index-{index}"));
    let raw_role = raw_value(object.get("role"));
    let role = normalize_role(&raw_role);
    let ignored = object
        .get("ignored")
        .and_then(Value::as_bool)
        .unwrap_or(false);
    let mut node = SnapshotNode {
        id,
        parent_id: raw_string_id(object.get("parentId")).unwrap_or_default(),
        role: role.clone(),
        name: raw_value(object.get("name")),
        value: raw_value(object.get("value")),
        url: String::new(),
        ignored,
        interactive: false,
        iframe: object
            .get("isFrameOwner")
            .and_then(Value::as_bool)
            .unwrap_or(false)
            || role.contains("iframe")
            || role == "frame"
            || raw_role.to_ascii_lowercase().contains("frame"),
        shadow_root: object
            .get("isShadowRoot")
            .and_then(Value::as_bool)
            .unwrap_or(false)
            || object
                .get("shadowRoot")
                .and_then(Value::as_bool)
                .unwrap_or(false)
            || object
                .get("shadowBoundary")
                .and_then(Value::as_bool)
                .unwrap_or(false)
            || role.contains("shadow"),
        backend_node_id: object
            .get("backendDOMNodeId")
            .and_then(raw_i64)
            .unwrap_or_default(),
        state: BTreeMap::new(),
        attributes: BTreeMap::new(),
        child_ids: object
            .get("childIds")
            .and_then(Value::as_array)
            .map(|items| {
                items
                    .iter()
                    .filter_map(|item| raw_string_id(Some(item)))
                    .collect()
            })
            .unwrap_or_default(),
        children: Vec::new(),
        visible: object
            .get("visible")
            .and_then(Value::as_bool)
            .unwrap_or(!ignored),
        dom_path: String::new(),
        sibling_ordinal: 0,
    };
    if object.get("frameId").and_then(Value::as_str).is_some()
        && (role == "iframe" || role == "frame")
    {
        node.iframe = true;
    }
    if let Some(attributes) = object.get("attributes").and_then(Value::as_object) {
        for (key, value) in attributes {
            node.attributes
                .insert(key.trim().to_ascii_lowercase(), raw_value(Some(value)));
        }
    }
    for key in ["alt", "title", "placeholder", "testid"] {
        if let Some(value) = object
            .get(key)
            .and_then(Value::as_str)
            .filter(|v| !v.is_empty())
        {
            node.attributes.insert(key.to_owned(), value.to_owned());
        }
    }
    if let Some(properties) = object.get("properties").and_then(Value::as_array) {
        for property in properties {
            let Some(property) = property.as_object() else {
                continue;
            };
            let name = property
                .get("name")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .trim()
                .to_ascii_lowercase();
            let value = raw_value(property.get("value"));
            match name.as_str() {
                "url" | "href" => {
                    if node.url.is_empty() {
                        node.url = value;
                    }
                }
                "focusable" | "editable" | "checked" | "expanded" | "selected" | "disabled"
                | "pressed" => {
                    if (name != "focusable" && name != "editable")
                        || raw_bool(property.get("value"))
                    {
                        node.state.insert(name.clone(), value);
                    }
                    if raw_bool(property.get("value")) && name != "expanded" {
                        node.interactive = true;
                    }
                }
                "type" | "autocomplete" if !value.is_empty() => {
                    node.attributes.insert(name, value);
                }
                _ => {}
            }
        }
    }
    if node.role == "link" && node.url.is_empty() {
        node.url = object
            .get("url")
            .or_else(|| object.get("href"))
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned();
    }
    node.interactive |= interactive_role(&node.role);
    Ok(node)
}

#[allow(clippy::too_many_arguments)]
fn render_node(
    parsed: &BTreeMap<String, SnapshotNode>,
    id: &str,
    options: &SnapshotOptions,
    depth: usize,
    visited: &mut BTreeSet<String>,
    ref_number: &mut usize,
    refs: &mut BTreeMap<String, SnapshotRef>,
    lines: &mut Vec<String>,
) {
    let Some(node) = parsed.get(id) else { return };
    if visited.contains(id) || (options.depth > 0 && depth as i32 > options.depth) {
        return;
    }
    visited.insert(id.to_owned());
    let include = !node.ignored
        && (!options.interactive || node.interactive || node.iframe || node.shadow_root)
        && (!options.compact || node.interactive || node.iframe || node.shadow_root);
    let mut children_depth = depth;
    if include {
        *ref_number += 1;
        let ref_name = format!("e{ref_number}");
        refs.insert(
            ref_name.clone(),
            SnapshotRef {
                node_id: node.id.clone(),
                backend_node_id: node.backend_node_id,
                role: node.role.clone(),
                name: node.name.clone(),
                value: node.value.clone(),
                state: snapshot_state(&node.state),
                visible: node.visible,
                interactive: node.interactive,
                url: node.url.clone(),
                refkey: ref_key(
                    &node.role,
                    &node.name,
                    &node.dom_path,
                    node.sibling_ordinal as i32,
                ),
                dom_path: node.dom_path.clone(),
                sibling_ordinal: node.sibling_ordinal,
                attributes: node.attributes.clone(),
            },
        );
        let mut line = format!("{}- {}", "  ".repeat(depth), node.role);
        if !node.name.is_empty() {
            line.push_str(&format!(" \"{}\"", node.name.replace('"', "\\\"")));
        } else if !node.value.is_empty() {
            line.push_str(&format!(" = \"{}\"", node.value.replace('"', "\\\"")));
        }
        if node.iframe {
            line.push_str(" [iframe]");
        }
        if node.shadow_root {
            line.push_str(" [shadow-root]");
        }
        if options.urls && !node.url.is_empty() {
            line.push_str(&format!(" ({})", node.url));
        }
        line.push_str(&format!(" [ref={ref_name}]"));
        lines.push(line);
        children_depth += 1;
    }
    for child in &node.children {
        render_node(
            parsed,
            child,
            options,
            children_depth,
            visited,
            ref_number,
            refs,
            lines,
        );
    }
}

fn assign_paths(parsed: &mut BTreeMap<String, SnapshotNode>, roots: &[String]) {
    for (ordinal, root_id) in roots.iter().enumerate() {
        if let Some(node) = parsed.get_mut(root_id) {
            node.sibling_ordinal = ordinal;
            let path = format!("/{}", path_segment(node));
            assign_path(parsed, root_id, path);
        }
    }
}

fn assign_path(parsed: &mut BTreeMap<String, SnapshotNode>, id: &str, path: String) {
    let Some(node) = parsed.get_mut(id) else {
        return;
    };
    node.dom_path = path.clone();
    let mut children = node.children.clone();
    children.sort_by(|left, right| {
        let left_key = parsed
            .get(left)
            .map(|child| {
                format!(
                    "{}{{{}}}",
                    path_segment(child),
                    subtree_signature(parsed, left, &mut BTreeSet::new())
                )
            })
            .unwrap_or_else(|| left.clone());
        let right_key = parsed
            .get(right)
            .map(|child| {
                format!(
                    "{}{{{}}}",
                    path_segment(child),
                    subtree_signature(parsed, right, &mut BTreeSet::new())
                )
            })
            .unwrap_or_else(|| right.clone());
        left_key.cmp(&right_key).then_with(|| left.cmp(right))
    });
    for (ordinal, child_id) in children.iter().enumerate() {
        if let Some(child) = parsed.get_mut(child_id) {
            child.sibling_ordinal = ordinal;
            let child_path = format!("{path}/{}", path_segment(child));
            assign_path(parsed, child_id, child_path);
        }
    }
}

fn subtree_signature(
    parsed: &BTreeMap<String, SnapshotNode>,
    id: &str,
    visiting: &mut BTreeSet<String>,
) -> String {
    if !visiting.insert(id.to_owned()) {
        return "cycle".to_owned();
    }
    let Some(node) = parsed.get(id) else {
        return "missing".to_owned();
    };
    let mut children: Vec<String> = node
        .children
        .iter()
        .map(|child| subtree_signature(parsed, child, visiting))
        .collect();
    children.sort();
    visiting.remove(id);
    format!("{}({})", path_segment(node), children.join(","))
}

fn path_segment(node: &SnapshotNode) -> String {
    let mut segment = normalize_ref_part(&node.role);
    if !node.name.is_empty() {
        segment.push('[');
        segment.push_str(&normalize_ref_part(&node.name));
        segment.push(']');
    }
    segment
}

fn raw_value(value: Option<&Value>) -> String {
    let Some(value) = value else {
        return String::new();
    };
    if value.is_null() {
        return String::new();
    }
    if let Some(inner) = value.as_object().and_then(|object| object.get("value")) {
        return raw_scalar(inner);
    }
    raw_scalar(value)
}

fn raw_scalar(value: &Value) -> String {
    match value {
        Value::Null => String::new(),
        Value::String(value) => value.clone(),
        Value::Bool(value) => value.to_string(),
        Value::Number(value) => value.to_string(),
        Value::Array(_) | Value::Object(_) => value.to_string(),
    }
}

fn raw_string_id(value: Option<&Value>) -> Option<String> {
    let value = raw_value(value);
    (!value.is_empty()).then_some(value)
}

fn raw_i64(value: &Value) -> Option<i64> {
    raw_value(Some(value)).parse().ok()
}

fn raw_bool(value: Option<&Value>) -> bool {
    raw_value(value).eq_ignore_ascii_case("true")
}

fn normalize_role(role: &str) -> String {
    let role = role.trim();
    if role.eq_ignore_ascii_case("rootwebarea") || role.eq_ignore_ascii_case("webarea") {
        "document".to_owned()
    } else if role.is_empty() {
        "generic".to_owned()
    } else {
        role.to_ascii_lowercase()
    }
}

fn interactive_role(role: &str) -> bool {
    matches!(
        role,
        "button"
            | "checkbox"
            | "combobox"
            | "link"
            | "listbox"
            | "menuitem"
            | "option"
            | "radio"
            | "searchbox"
            | "slider"
            | "spinbutton"
            | "switch"
            | "tab"
            | "textbox"
            | "treeitem"
            | "gridcell"
    )
}

fn snapshot_state(state: &BTreeMap<String, String>) -> String {
    state
        .iter()
        .map(|(key, value)| format!("{key}={value}"))
        .collect::<Vec<_>>()
        .join(",")
}

fn snapshot_id_cmp(left: &str, right: &str) -> Ordering {
    match (left.parse::<i64>(), right.parse::<i64>()) {
        (Ok(left), Ok(right)) if left != right => left.cmp(&right),
        _ => left.cmp(right),
    }
}

fn is_zero(value: &usize) -> bool {
    *value == 0
}
fn is_zero_i32(value: &i32) -> bool {
    *value == 0
}
fn is_zero_i64(value: &i64) -> bool {
    *value == 0
}
